File: exercise/lib/block/extract_common_behaviour.rb
require "socket"

class Client
  def initialize(ip="127.0.0.1",port=3333)
    @ip, @port = ip, port
  end

  def connect(&block)
    socket = TCPSocket.new(@ip,@port)
    yield socket
  ensure
    socket.close
  end

  def send_message(msg)
    connect do |socket|
      socket.puts(msg)
    end
  end

  def receive_message
    connect do |socket|
      socket.gets
    end
  end
end

File: examples/lib/block/java_style/extract_common_behaviour.java
require "socket"

class Client {
  private String ip;
  private Integer port;

  public Client(String ip, Integer port) {
    this.ip = ip;
    this.port = port;
  }

  public void sendMessage(String msg) {
    TCPSocket socket = TCPSocket.new(this.ip, this.port);
    try {
      socket.puts(msg);
    } finally {
      socket.close();
    }
  }

  public String receiveMessage {
    TCPSocket socket = TCPSocket.new(this.ip, this.port);
    try {
      return socket.gets();
    } finally {
      socket.close();
    }
  }
}
